[export] add translations for SymInt/Bool deserialization; FloorDiv

pianpwk opened the pull request:
Summary:
fix for https://github.com/pytorch/pytorch/issues/136797

We have an issue with deserialization for symbolic expressions, where the original expression has a structure mismatch with the deserialized expression. This happens because we serialize the expressions into strings, and deserialize using sympy.sympify().

This seems to cause issues with custom torch sympy.Functions, in this case FloorDiv. In this case, the expression `FloorDiv(a, b)` turns into `sympy.floor(a, PowByNatural(b, -1))`, or `sympy.floor(sympy.Mul(*a), PowByNatural(b, -1))` if a is also a sympy.Mul. This seems to break downstream when we call torch.empty_strided() on this expression, and the PowByNatural has an invalid ValueRange (e.g. [0, -1]).

This fixes that by adding translations after deserialization, to pattern match and transform expressions into the original format. Other potential strategies are a) we could also serialize SymInts as sympy.Expr structures, but that seems a lot more involved and BC breaking, or b) we could modify sympy.sympify(?), but I don't know if this is possible.

Test Plan: test_export/test_serialize

Differential Revision: D63493615

pianpwk commented:
@ezyang would you know of a better way to do this? It seems like even if we let the invalid ValueRange pass or flip min/max, PowByNatural + ValueRanges is the thing to avoid, because it still means we lose precision when truncating the range to ints (e.g. [0, -1]).

angelayi commented on torch/_export/serde/serialize.py:
```diff
@@ -1540,6 +1569,7 @@ def deserialize_sym_int(self, s: SymInt) -> Union[int, torch.SymInt]:
                                 compiler_max=vr.upper,  # type: ignore[arg-type]
                             )
 
+            sym = _translate_deserialized_sym_expr(sym)
```
should this line go directly after line 1526?

pianpwk replied:
updated

angelayi commented on torch/_export/serde/serialize.py:
```diff
@@ -1562,6 +1592,7 @@ def deserialize_sym_bool(self, s: SymBool) -> Union[bool, torch.SymBool]:
                     self.deserialize_sym_int(SymInt.create(as_expr=SymExpr(str(sym))))
             # then we sympify again using locals to correctly reify with the constructed symbols
             expr = sympy.sympify(val.expr_str, locals=self.symbol_name_to_symbol)
+            expr = _translate_deserialized_sym_expr(expr)  # is this necessary?
```
idk if this is necessary... TrueDiv doesnt act on symbools right..?

ezyang commented:
So, the proper way to fix this is to stop serializing the Sympy string expressions as strings, for which we have NO stability or even round-trippability guarantee. Sympy expressions are supposed to be roundtrippable to FX IR, and back, via the Python reference / sympy interpreter stuff (there are some bugs / incompleteness, but we should fix those anyway.)

But I'm guessing you didn't want to hear that? Lol.

pianpwk commented:
> So, the proper way to fix this is to stop serializing the Sympy string expressions as strings, for which we have NO stability or even round-trippability guarantee. Sympy expressions are supposed to be roundtrippable to FX IR, and back, via the Python reference / sympy interpreter stuff (there are some bugs / incompleteness, but we should fix those anyway.)
> 
> But I'm guessing you didn't want to hear that? Lol.

Hahaha no but this makes total sense

ezyang commented:
If you are dead set on string format, I'd probably still want to have a customized printer and parser. In particular the parser needs to override the mapping of FloorDiv to which class. As a stopgap, we could potentially rename functions in torch.utils._sympy.functions so they don't conflict with Sympy

bhack commented:
I think we had the same problem with inductor cache right?

 ```python
site-packages/torch/utils/_sympy/interp.py:159] [0/1] failed while executing pow_by_natural([VR[0, int_oo], VR[-1, -1]])
site-packages/torch/_inductor/codecache.py:609] [0/1] Can't pickle
site-packages/torch/_inductor/codecache.py:609] [0/1] Traceback (most recent call last):
site-packages/torch/_inductor/codecache.py:609] [0/1]   File "/opt/conda/lib/python3.11/site-packages/torch/_inductor/codecache.py", line 605, in dumps
site-packages/torch/_inductor/codecache.py:609] [0/1]     pickler.dump(obj)
site-packages/torch/_inductor/codecache.py:609] [0/1] AttributeError: Can't pickle local object 'make_opaque_unary_fn.<locals>.OpaqueUnaryFn'
site-packages/torch/utils/_sympy/interp.py:159] [0/1] failed while executing pow_by_natural([VR[0, int_oo], VR[-1, -1]])
```

pianpwk committed:
[export] add translations for SymInt/Bool deserialization; FloorDiv (#136802)

Summary:

fix for https://github.com/pytorch/pytorch/issues/136797

We have an issue with deserialization for symbolic expressions, where the original expression has a structure mismatch with the deserialized expression. This happens because we serialize the expressions into strings, and deserialize using sympy.sympify().

This seems to cause issues with custom torch sympy.Functions, in this case FloorDiv. In this case, this alters the structure from `FloorDiv(a, b)` into `sympy.floor(a, PowByNatural(b, -1))`, or `sympy.floor(sympy.Mul(*a), PowByNatural(b, -1))` if a is also a sympy.Mul. This seems to break downstream when we call torch.empty_strided() on this expression, and the PowByNatural has an invalid ValueRange (e.g. [0, -1]).

This fixes that by adding translations after deserialization, to pattern match and transform expressions into the original format. Other potential strategies are a) we could also serialize SymInts as sympy.Expr structures, but that seems a lot more involved and BC breaking, or b) we could modify sympy.sympify(?), but I don't know if this is possible.

Test Plan: test_export/test_serialize

Differential Revision: D63493615

pianpwk commented:
Yeah, I think we should just move to not string format, and add to the export serialization schema

bhack commented:
> https://github.com/pytorch/pytorch/pull/136802#issuecomment-2384520332

Is this related right?

bhack commented:
Is there a roadmap to review this?

bhack commented:
Is this stalled?

bhack commented:
I've tried to hotpatch the latest nightly with this PR and the deserializzation issue it is still here.

bhack commented:
if this is going to be close as stale do we still have a ticket to track this?

avikchaudhuri commented:
Was this fixed by https://github.com/pytorch/pytorch/pull/141284?

bhack commented:
I've tested on Nov, 2  and it was merge on Nov, 22. So probably it is ok
